### PR TITLE
Feature/questions answers/tutor approval

### PR DESCRIPTION
**Pull request by filip-gregurevic**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] ~~Server: I added multiple integration tests (Spring) related to the features~~
- [x] ~~Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)~~
- [x] ~~Server: I implemented the changes with a good performance and prevented too many database calls~~
- [x] ~~Server: I documented the Java code using JavaDoc style.~~
- [x]  Client: I added multiple integration tests (Jest) related to the features
- [x] ~~Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)~~
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

There is no way of indicating that an answer to a question is correct except explicitly writing it. Students have to read through all questions in order to find if there is a correct answer or which one it is.

### Description

Tutors and instructors can mark an answer as approved with an additional button. Approved answers are highlighted and listed first. Unapproved answers are hidden when loaded and can be shown by clicking **show more answers**. They can be hidden again by clicking **show less answers**. If there are no approved answers all answers are shown. If there are only approved answers, the option to show more / less disappears.

### Steps for Testing

1. Log in to Artemis
(1.a Set up a course in which students can post questions / answers)
2. Navigate to an exercise or a lecture
3. Create a question
4. Create multiple answers
4.a Mark an answer as approved
4.b Click 'show more answers'
4.c Click 'show less answers'
4.d Unmark an answer as approved
5. Mark all answers as approved
6. Unmark all answers as approved

### Screenshots

1. New button for tutors and instructors to approve answers
Bildschirmfoto 2020-03-31 um 09 23 47

2. Approved answers are highlighted
Bildschirmfoto 2020-04-01 um 11 12 50

3. If no answers are approved all answers are shown
Bildschirmfoto 2020-04-01 um 11 12 41

4. If all answers are approved the option to show more answers disappears 
Bildschirmfoto 2020-04-01 um 11 13 10

5. More answers can be expanded
Bildschirmfoto 2020-04-01 um 11 12 57

### CODE REVIEW

File: src/main/webapp/app/overview/student-questions/student-question-row.component.ts [hunk 1]
```diff
@@ -115,10 +144,25 @@ export class StudentQuestionRowComponent implements OnInit, OnDestroy {
         });
     }
 
+    /**
+     * Takes a studentAnswer and deletes it
+     * @param   {studentQuestionAnswer} studentAnser
```
Comment (sascha11110) [inline src/main/webapp/app/overview/student-questions/student-question-row.component.ts:149]: `studentAnser` little typo here :-)

File: src/main/webapp/app/overview/student-questions/student-question-row.component.ts [hunk 2]
```diff
@@ -115,10 +144,25 @@ export class StudentQuestionRowComponent implements OnInit, OnDestroy {
...
     deleteAnswer(studentAnswer: StudentQuestionAnswer): void {
         this.studentQuestionAnswerService.delete(studentAnswer.id).subscribe((res: HttpResponse<any>) => {
             this.studentQuestion.answers = this.studentQuestion.answers.filter((el) => el.id !== studentAnswer.id);
             this.sortQuestionAnswers();
         });
     }
+
+    /**
+     * Takes a studentAnswer and toggles the tutorApproved field
+     * @param   {studentQuestionAnswer} studentAnser
```
Comment (sascha11110) [inline src/main/webapp/app/overview/student-questions/student-question-row.component.ts:160]: Same typo

File: src/main/webapp/app/overview/student-questions/student-question-row.component.ts (hunk 1 above)
Comment (filip-gregurevic) [reply src/main/webapp/app/overview/student-questions/student-question-row.component.ts]: Thanks! Fixed this typo

File: src/main/webapp/app/overview/student-questions/student-question-row.component.ts (hunk 2 above)
Comment (filip-gregurevic) [reply src/main/webapp/app/overview/student-questions/student-question-row.component.ts]: Fixed too :)

**Comment by filip-gregurevic** [comment]
> Nice work! Very intuitive to use! Code-wise everything looks ok to me.
> 
> One server-side issue that I found when testing:
> 
> 1. When you add an answer and then try to update it, you will receive a 500 server error due to a `NullPointerException`. The cause is that `updateStudentQuestionAnswer` uses `studentQuestionAnswer` from the request body to perform the `mayUpdateOrDeleteStudentQuestionAnswer` check. The course nested on the question is missing in the payload though, thus causing a `NullPointerException`.
> 
> Small client-side bugs I found:
> 
> 1. When editing an anwer and then clicking on the edit icon of another answer, the latter answer disappears but no edit input is shown.
> 2. When editing a question and saving it, an error gets thrown: `ERROR TypeError: t.creationDate.isValid is not a function`
> 
> UI improvements that are more in the category of "nice to have":
> 
> 1. The "Show more/less answers" does not have a pointer cursor indicating that it is a clickable action. It's also not accessible via keyboard (same for all the other actions).
> 2. There is no validation. It is possible to save empty questions or answers.
> 3. Is the order of the questions on purpose? I think it might make sense to reverse it and show the newest question at the top so people directly notice.

@madwau thanks for the feedback, those are some good catches! I'll definitely have a look at them.

But are those existing bugs inside the scope of this PR?
Since I only added the _approved by a tutor_  functionality and didn't newly implement or edit any of the functionalities you mentioned, I'd rather merge it as it is functioning and fix the bugs in a separate PR.

**Comment by madwau** [comment]
> @madwau thanks for the feedback, those are some good catches! I'll definitely have a look at them.
> 
> But are those existing bugs inside the scope of this PR?
> Since I only added the _approved by a tutor_ functionality and didn't newly implement or edit any of the functionalities you mentioned, I'd rather merge it as it is functioning and fix the bugs in a separate PR.

I didn't check which of the bugs were actually caused by this PR, just tested the functionality in general.

Sure, solving them in a separate PR sounds good to me as well :)

Maybe you can fix the one server-side bug here since it's a little more critical and very easy to resolve?

**Comment by filip-gregurevic** [comment]
> > @madwau thanks for the feedback, those are some good catches! I'll definitely have a look at them.
> > But are those existing bugs inside the scope of this PR?
> > Since I only added the _approved by a tutor_ functionality and didn't newly implement or edit any of the functionalities you mentioned, I'd rather merge it as it is functioning and fix the bugs in a separate PR.
> 
> I didn't check which of the bugs were actually caused by this PR, just tested the functionality in general.
> 
> Sure, solving them in a separate PR sounds good to me as well :)
> 
> Maybe you can fix the one server-side bug here since it's a little more critical and very easy to resolve?

@madwau The server-error should be fixed now with https://github.com/ls1intum/Artemis/pull/1285/commits/7da4f30ea4115710e51dd5819ce80c7408cbb28c
Now it passes the studentQuestionAnswer from the db query.
I'll fix the other bugs in a separate PR then :)

**Comment by madwau** [comment]
> @madwau The server-error should be fixed now with [7da4f30](https://github.com/ls1intum/Artemis/commit/7da4f30ea4115710e51dd5819ce80c7408cbb28c)
> Now it passes the studentQuestionAnswer from the db query.
> I'll fix the other bugs in a separate PR then :)

Nice! Just verified and it works now, thanks.

**Pull request merged**